### PR TITLE
A generic workflow to build Sphinx documentation and deploy to gitHub Pages

### LINKED ISSUE #34: Optimise build-docs workflow
In #20 a workflow to implement sphinx documentation builds was created. However, this workflow makes use of pip to install dependencies and the speed could be improved by getting the workflow to install dependencies via uv.

### PR DESCRIPTION
**Pull request by yaswant**
# PR Summary

Code Reviewer: @james-bruten-mo 

closes #34 

A generic reusable workflow to build and deploy sphinx-based documentation with minimal effort.

## Code Quality Checklist

- [x] I have performed a self-review of my own code
- [x] My code follows the project's style guidelines
- [x] The modified workflow's README has been updated, if required
- [x] The changes have been sufficiently tested (see https://github.com/MetOffice/git_playground/pull/136)

## AI Assistance and Attribution

- [x] Some sections of the `sphinx-docs/README.md` were enhanced with assistance from Met Office Github Copilot Enterprise and I have followed the [Simulation Systems AI policy](https://metoffice.github.io/simulation-systems/FurtherDetails/ai.html)

# Code Review

- [ ] The changes are approriate and testing has been sufficient

### CODE REVIEW

File: .github/workflows/sphinx-docs.yaml [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+
+      - name: Build HTML docs (sphinx-build)
+        if: ${{ steps.detect-makefile.outputs.found == 'false' }}
+        run: |
+          sphinx-build \
+            --builder html \
+            --fresh-env \
+            --write-all \
+            "${{ inputs.source-dir }}" \
+            "${{ inputs.docs-dir }}/${{ inputs.build-dir }}/html"
```
Comment (andrewcoughtrie) [inline .github/workflows/sphinx-docs.yaml]: We should add `--nitpicky`, `--fail-on-warning`, `--keep-going` to make sure the build is clean the warnings in particular are important. The default ubuntu GitHub runners also have 4 cpus cores so adding `-j 4` or giving the number of cores as an option to the workflow would help with performance.

File: .github/workflows/sphinx-docs.yaml [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+        run: |
+          if [ -f "${{ inputs.docs-dir }}/Makefile" ]; then
+            echo "found=true" >> "$GITHUB_OUTPUT"
+            echo "Makefile detected in '${{ inputs.docs-dir }}'. Will use 'make html' to build docs."
+          else
+            echo "found=false" >> "$GITHUB_OUTPUT"
+            echo "No Makefile detected in '${{ inputs.docs-dir }}'. Will use 'sphinx-build' to build docs."
+          fi
+
+      - name: Build HTML docs (make)
```
Comment (andrewcoughtrie) [inline .github/workflows/sphinx-docs.yaml:299]: It would be useful to use the [`runner.debug` context](https://docs.github.com/en/actions/how-tos/monitor-workflows/enable-debug-logging#enabling-step-debug-logging) to add additional debug messaging. Not only for the script but for the Sphinx run as well, adding flags like `--verbose` and `--show-traceback` and possibly `--pdb` (the python debugger) to the sphinx command might make it easier to determine problems when re-running with the debug box ticked. Adding any debug options to UV and possibly extending the timeout time during a debug run might be helpful as well.

**Review by andrewcoughtrie — CHANGES_REQUESTED** [review]
Looks good on the whole, a couple of suggestions to help with debugging etc.

File: .github/workflows/sphinx-docs.yaml (hunk 1 above)
Comment (yaswant) [reply .github/workflows/sphinx-docs.yaml]: Thanks! Good suggestions. Included them all, with one exception - the 4 vCPUs are available for _public_ repositories, which typically default to 2vCPUs on _private_ repositories. `auto` is the safe option which will use number of available CPU cores.

File: .github/workflows/sphinx-docs.yaml (hunk 2 above)
Comment (yaswant) [reply .github/workflows/sphinx-docs.yaml:299]: That's added in https://github.com/MetOffice/growss/pull/76/commits/dcb50e9290b2fe5fc37e609b4816c3dce5bcd8ba although I am very reluctant to enable `pdb` here. Can I convince you with full traceback dump in verbose mode?

As I understand, we should not use `pdb` in non-interactive mode - this will unnecessarily keep the "standard" GitHub Action runs in a hanging state upon hitting a breakpoint.

**Review by andrewcoughtrie — DISMISSED** [review]
This all looks good to me now.

**Comment by andrewcoughtrie** [comment]
@james-bruten-mo do you want to have a look at this as well?

File: .github/workflows/sphinx-docs.yaml [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+        run: echo "${{ github.workspace }}/${{ env.VENV_PATH }}/bin" >> "$GITHUB_PATH"
+
+      - name: Lint Sphinx docs
+        run: sphinx-lint ${{ inputs.source-dir }}
+
+      - name: Detect Makefile in docs directory
+        id: detect-makefile
+        run: |
+          if [ -f "${{ inputs.docs-dir }}/Makefile" ]; then
+            echo "found=true" >> "$GITHUB_OUTPUT"
```
Comment (james-bruten-mo) [inline .github/workflows/sphinx-docs.yaml]: I suggest `found` should be something more explicit here, `found_makefile` if nothing else

**Review by james-bruten-mo — CHANGES_REQUESTED** [review]
I've got one minor change request to rename one of the variables pushed to github_env.

I guess I'm not completely convinced by this as a change. In the repos we manage, we'll be replacing a handful of relatively simple workflows with a single huge and complicated ones - I'm not sure that in this case, we're increasing maintainability? 
I also really don't like the use of python throughout. Effectively we've got inline python, within a bash script, within a yaml file, which is something that you can do, but I'm not convinced we should. I'm not suggesting that it be rewritten as bash, but maybe it'd be better storing in a separate python file and just calling that?

The workflow seems reasonably impressive in many ways - I just worry that it's a little too complicated to be increasing maintainability (not to mention if it does break, all repos will be affected rather than just one at the moment). If you want to keep pursuing, then sure, but just noting some concerns.

File: .github/workflows/sphinx-docs.yaml (hunk 3 above)
Comment (yaswant) [reply .github/workflows/sphinx-docs.yaml]: done.

**Comment by yaswant** [comment]
> I've got one minor change request to rename one of the variables pushed to github_env.
> 
> I guess I'm not completely convinced by this as a change. In the repos we manage, we'll be replacing a handful of relatively simple workflows with a single huge and complicated ones - I'm not sure that in this case, we're increasing maintainability? I also really don't like the use of python throughout. Effectively we've got inline python, within a bash script, within a yaml file, which is something that you can do, but I'm not convinced we should. I'm not suggesting that it be rewritten as bash, but maybe it'd be better storing in a separate python file and just calling that?
> 
> The workflow seems reasonably impressive in many ways - I just worry that it's a little too complicated to be increasing maintainability (not to mention if it does break, all repos will be affected rather than just one at the moment). If you want to keep pursuing, then sure, but just noting some concerns.

I understand your views and myself thought about keeping the scripts part separate (e.g., .github/scripts). But that will make the use of reusable workflow more complex when called from a forked PR unless we explicitly handle the checkout process . Keeping the messy python bit inline is guaranteed to work without additional work.

If you could test a few edge cases where you think this might fail, that would be great

**Comment by yaswant** [comment]
@james-bruten-mo thanks for the review and for some critical points James raised. It made me think again! So, I've made some changes to address some concerns:

1. The dependency resolution using `conf.py` was too long, too nested, a maintenance overhead. I've simplified this - if you are a fan of list comprehension you will appreciate the changes. This will resolve all dependencies we have in any SimIT documentation repositories. 
  However, separating this as a standalone script will be counter-productive and will require checking out the script explicitly (unnecessary overhead + dependency).
1. Replaced the toml parsing from python to [`yq`](https://mikefarah.gitbook.io/) (for yaml, but can handle toml, one-way) - I thought it was not available on action runner. Turns out its now available on ubuntu-* and macos runners. So if you are happy with the changes, I'll update the README to indicate that this is not 100% compatible on Windows runners or add an extra step to install if runner is Windows (for the sake of completeness)..

@andrewcoughtrie thanks for the discussion around ubuntu-slim runner offline. This is a perfect choice for sphinx-build on our private repositories.


I've tested few configurations here: <https://github.com/MetOffice/git_playground/pull/136>

**Review by james-bruten-mo — APPROVED** [review]
Thanks Yash, that looks better with the reduced python. Do you want to update the readme - I don't think we need to worry about windows runner support

**Review by andrewcoughtrie — APPROVED** [review]

**Comment by yaswant** [comment]
@james-bruten-mo I have now updated the workflow to avoid code injection via variable expansion + updated the README to warn incompatibility on Windows runners.

Tests in: https://github.com/MetOffice/git_playground/pull/136

**Review by james-bruten-mo — APPROVED** [review]
Looks good to me
